bazarr/subtitles/upgrade.py

```diff
@@ -270,31 +282,78 @@ def get_upgradable_episode_subtitles():
+               TableHistory.upgradedFromId)
+        .select_from(TableHistory)
+        .join(TableShows, onclause=TableHistory.sonarrSeriesId == TableShows.sonarrSeriesId)
+        .join(TableEpisodes, onclause=TableHistory.sonarrEpisodeId == TableEpisodes.sonarrEpisodeId)
         .join(max_id_timestamp, onclause=and_(TableHistory.video_path == max_id_timestamp.c.video_path,
                                               TableHistory.language == max_id_timestamp.c.language,
-                                              max_id_timestamp.c.timestamp == TableHistory.timestamp)) \
-        .join(TableShows, onclause=TableHistory.sonarrSeriesId == TableShows.sonarrSeriesId) \
-        .join(TableEpisodes, onclause=TableHistory.sonarrEpisodeId == TableEpisodes.sonarrEpisodeId) \
-        .where(reduce(operator.and_, upgradable_episodes_conditions)) \
-        .order_by(TableHistory.timestamp.desc())\
-        .subquery()
+                                              max_id_timestamp.c.timestamp == TableHistory.timestamp))
+        .where(reduce(operator.and_, upgradable_episodes_conditions))
+        .order_by(TableHistory.timestamp.desc())) \
+        .all()
+    logging.debug(f"{len(subtitles_to_upgrade)} subtitles are candidates and we've selected the latest timestamp for "
+                  f"each of them.")
+
+    query_actions_without_upgrade = [x for x in query_actions if x != 3]
+    upgradable_episode_subtitles = {}
+    for subtitle_to_upgrade in subtitles_to_upgrade:
+        # check if we have the original subtitles id in database and use it instead of guessing
+        if subtitle_to_upgrade.upgradedFromId:
+            upgradable_episode_subtitles.update({subtitle_to_upgrade.id: subtitle_to_upgrade.upgradedFromId})
+            logging.debug(f"The original subtitles ID for TableHistory ID {subtitle_to_upgrade.id} stored in DB is: "
+                          f"{subtitle_to_upgrade.upgradedFromId}")
+            continue
+
+        # if not, we have to try to guess the original subtitles id
+        logging.debug("We don't have the original subtitles ID for this subtitle so we'll have to guess it.")
+        potential_parents = database.execute(
+            select(TableHistory.id, TableHistory.action)
+            .where(TableHistory.video_path == subtitle_to_upgrade.video_path,
+                   TableHistory.language == subtitle_to_upgrade.language,)
+            .order_by(TableHistory.timestamp.desc())
+        ).all()
+
+        logging.debug(f"The potential original subtitles IDs for TableHistory ID {subtitle_to_upgrade.id} are: "
+                      f"{[x.id for x in potential_parents]}")
+        confirmed_parent = None
+        for potential_parent in potential_parents:
+            if potential_parent.action in query_actions_without_upgrade:
+                confirmed_parent = potential_parent.id
+                logging.debug(f"This ID is the first one to match selected query actions so it's been selected as "
```

Comment:
In what sense is this one the "first" to match?
Is it the first oldest one or the first newest one?